model/generators/mento.py

```diff
@@ -0,0 +1,169 @@
+from model.utils.generator import Generator, state_update_blocks
+from model.utils import update_from_signal
+
+# raise numpy warnings as errors
+np.seterr(all='raise')
+
+class MentoExchangeGenerator(Generator):
+    """
+    The MentoExchangeGenerator emulates Mento AMMs that
+    function as cross-product automated market maker between
+    a virtual stable bucket and a reserve currency bucket.
+    In v1 that's one of cUSD, cEUR and cREAL with CELO as
+    the reserve currency.
+    But the generator permits modeling AMMs backed by
+    any reserves on other chains as well for example (cUSD, ETH).
+    """
+    configs: Dict[MentoExchange, MentoExchangeConfig]
+    active_exchanges: Set[MentoExchange]
+
+    def __init__(self, configs: Dict[Stable, MentoExchangeConfig], active_exchanges: Set[Stable]):
+        self.configs = configs
+        self.active_exchanges = active_exchanges
+
+    @classmethod
+    def from_parameters(cls, params, _initial_state, _container):
+        return cls(
+            params['mento_exchanges_config'],
+            set(params['mento_exchanges_active'])
+        )
+
+    @state_update_blocks('bucket_update')
+    def bucket_update(self):
+        return [{
+            'description': """
+                Updates blocks only when update bucket_update_frequency_seconds has passed since last update.
+            """,
+            'policies': {
+                'bucket_update': self.get_bucket_update_policy()
+            },
+            'variables': {
+                'mento_buckets': update_from_signal('mento_buckets')
+            }
+        }]
+
+    def get_bucket_update_policy(self):
+        """
+        Policy function which updates AMM buckets
+        """
+        def p_bucket_update(
+            _params,
+            _substep,
+            _state_history,
+            prev_state,
+        ):
+            mento_buckets = {
+                exchange: self.get_next_buckets(prev_state, exchange)
+                for exchange in self.active_exchanges
+            }
+
+            return {
+                'mento_buckets': mento_buckets
+            }
+        return p_bucket_update
+
+    def get_next_buckets(self, prev_state, exchange):
+        """
+        Get the next bucket sizes for a give exchange
+        """
+        update_required = self.buckets_should_be_reset(exchange, prev_state)
+        if update_required:
+            return self.recalculate_buckets(exchange, prev_state)
+        return prev_state['mento_buckets'][exchange]
+
+
+    def buckets_should_be_reset(self, exchange: MentoExchange, prev_state) -> bool:
+        """
+        Returns true if the buckets for a particular exchange have to be reset
+        """
+        bucket_update_frequency = self.configs[exchange].bucket_update_frequency_second
+        update_required = (
+            (blocktime_seconds * prev_state['timestep']) % bucket_update_frequency == 0
+        ) or (prev_state['timestep'] == 1)
```

Comment:
It was because the bucket sizes are dependent on the system parameters and those are dependent on the parameter sweep so there's no clear way to do that "outside" of the simulation besides adding an extra step to the `Engine`. 
We could have something like if the value of a state variable is a lambda, we call it with the parameters for that run. We could add an issue for that, but then we decided it's easier like this.